Jetchat/app/src/main/java/com/example/compose/jetchat/conversation/UserInput.kt

```diff
@@ -390,49 +417,151 @@ private fun UserInputText(
-                        style = MaterialTheme.typography.bodyLarge.copy(color = disableContentColor)
+        AnimatedContent(
+            targetState = isRecordingMessage,
+            label = "text-field",
+            modifier = Modifier
+                .weight(1f)
+                .fillMaxHeight()
+        ) { recording ->
+            Box(Modifier.fillMaxSize()) {
+                if (recording) {
+                    RecordingIndicator(swipeOffset)
+                } else {
+                    UserInputTextField(
+                        textFieldValue,
+                        onTextChanged,
+                        onTextFieldFocused,
+                        keyboardType,
+                        focusState
                     )
                 }
             }
         }
+        RecordButton(
+            recording = isRecordingMessage,
+            swipeOffset = swipeOffset,
+            onStartRecording = {
+                val consumed = !isRecordingMessage
+                isRecordingMessage = true
+                consumed
+            },
+            onFinishRecording = {
+                // handle end of recording
+                isRecordingMessage = false
+            },
+            onCancelRecording = {
+                isRecordingMessage = false
+            },
+            modifier = Modifier.fillMaxHeight()
+        )
+    }
+}
+
+@Composable
+private fun BoxScope.UserInputTextField(
+    textFieldValue: TextFieldValue,
+    onTextChanged: (TextFieldValue) -> Unit,
+    onTextFieldFocused: (Boolean) -> Unit,
+    keyboardType: KeyboardType,
+    focusState: Boolean
+) {
+    var lastFocusState by remember { mutableStateOf(false) }
+    BasicTextField(
+        value = textFieldValue,
+        onValueChange = { onTextChanged(it) },
+        modifier = Modifier
+            .padding(start = 32.dp)
+            .align(Alignment.CenterStart)
+            .onFocusChanged { state ->
+                if (lastFocusState != state.isFocused) {
+                    onTextFieldFocused(state.isFocused)
+                }
+                lastFocusState = state.isFocused
+            },
+        keyboardOptions = KeyboardOptions(
+            keyboardType = keyboardType,
+            imeAction = ImeAction.Send
+        ),
+        maxLines = 1,
+        cursorBrush = SolidColor(LocalContentColor.current),
+        textStyle = LocalTextStyle.current.copy(color = LocalContentColor.current)
+    )
+
+    val disableContentColor =
+        MaterialTheme.colorScheme.onSurfaceVariant
+    if (textFieldValue.text.isEmpty() && !focusState) {
+        Text(
+            modifier = Modifier
+                .align(Alignment.CenterStart)
+                .padding(start = 32.dp),
+            text = stringResource(R.string.textfield_hint),
+            style = MaterialTheme.typography.bodyLarge.copy(color = disableContentColor)
+        )
+    }
+}
+
+@Composable
+private fun RecordingIndicator(swipeOffset: State<Float>) {
```

Comment:
`swipeOffset` should be a lambda here not State. The pattern exists to make testing and previews easier to write because you don't need to alloc a state object and can just pass in something like `{ 0.5f }`